### PR TITLE
Expand docs on caller requirements around ids

### LINKED ISSUE #218: Update documentation on calling application requirments around session IDs
Once `Participant`s only exist for the duration of a single session, they're no longer able to check global requirements, like whether a new protocol request is using a unique session identifier. This becomes the responsibility of the calling application. We need to document the requirement that session IDs are globally unique -- e.g. that the calling application has never executed a protocol with the same session ID.

This is probably going to be one of several requirements on the calling application, so I'd recommend starting a section in the top-level library docs for this content or for a link to this content if you decide to put it elsewhere.

## Closing criteria
- [ ] Write the docs
- [ ] Make sure they're linked to or in the top-level library docs

### PR DESCRIPTION
**Pull request by marsella**
Closes #218, with a little extra on participant identifiers.

Requesting @Ra1issa as a reviewer since you have thought the most about session IDs.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -48,26 +48,43 @@
...
+//! agree on the set of identifiers. They must satisfy several properties:
+//!     1. All identifiers must be consistent across all
+//! participants in a session.
+//!     2. The session [`Identifier`] must be global and unique;
+//! in particular, it must not be reused across multiple protocol
+//! instances.
+//!     3. The [`ParticipantIdentifier`]s can be reused across multiple
+//! sessions, but we don't recommend reusing a single [`ParticipantIdentifier`]
+//! for different real-world entities.
+//!
```
Comment (indomitableSwan) [inline src/lib.rs:87]: Should we make this advice stronger? Is there a reason we don't just say that you shouldn't reuse a `ParticipantIdentifier` for different real-world entities? I'm not sure I understand the fallout from this type of identifier collision, but it seems like we're better off saying we don't make any guarantees about system behavior for this usage.

File: src/protocol.rs [hunk 2]
```diff
@@ -287,7 +287,18 @@ impl ParticipantConfig {
...
+/// All [`Participant`]s in a session must agree on the
+/// [`ParticipantIdentifier`]s. That is, these are not local identifiers for a
+/// single [`Participant`]; they are global identifiers for the [`Participant`]s
+/// in a session.
+///
+/// [`ParticipantIdentifier`]s can be used across multiple sessions. For
+/// example, if a set of participants run keygen, auxinfo, and then compute
+/// several signatures, they can use the same set of identifiers for each of
+/// those sessions. However, we do not recommend reusing identifiers across
+/// different participant sets or for different entities.
```
Comment (indomitableSwan) [inline src/protocol.rs:301]: Are participant identifiers meant to tie into how the calling application realizes message authentication? (An unstated assumption on networking is that the participants are connected via authenticated point-to-point channels and have access to an authenticated broadcast mechanism.)

File: src/protocol.rs [hunk 3]
```diff
@@ -313,42 +324,60 @@ impl std::fmt::Display for ParticipantIdentifier {
...
-/// parameters, proofs and commitments with their corresponding sessions.
+/// Session identifiers have two roles in the protocol: they tag messages
+/// and they are incorporated as context into zero-knowledge proofs.
+/// The "globally unique" property allows participants to distinguish
+/// messages belonging to different, concurrent protocol runs,
+/// prevents collisions between messages belonging to
+/// different sessions, and prevents replay attacks by associating messages and
+/// zero-knowledge proofs
+/// with the session, fixed parameters, and previous subprotocols to which
+/// they correspond.
```
Comment (indomitableSwan) [inline src/protocol.rs:338]: I would add that they are a requirement for achieving UC security.

Comment (indomitableSwan) [reply src/protocol.rs:338]: I would also clarify here that the calling application is responsible for picking session identifiers and what the requirements are. This is mentioned with some detail with respect to `ParticipantIdentifiers`, but for session identifiers, we seem to just say (below) that we expect them to be chosen uniformly at random. (But by who? And why are we dictating how they are chosen in this fashion?)

File: src/protocol.rs [hunk 4]
```diff
@@ -313,42 +324,60 @@ impl std::fmt::Display for ParticipantIdentifier {
...
-/// relying instead on a single type for all session identifiers. The
-/// distinction between sessions and sub-sessions is inherently enforced by the
-/// order of inputs and outputs to different stages in the protocol.
+/// 2. The paper distinguishes between sessions (keygen) and sub-sessions
+/// (auxinfo and presign).
+/// The codebase requires the calling application to select a new, unique
+/// session [`Identifier`]s at three points:
+/// (1) immediately before starting a new keygen session;
+/// (2) immediately before starting a new auxinfo session;
+/// (3) immediately before starting a new presigning session.
```
Comment (indomitableSwan) [inline src/protocol.rs:360]: The session identifier for signing should be used both for presigning and for signing steps.

File: src/protocol.rs [hunk 5]
```diff
@@ -313,42 +324,60 @@ impl std::fmt::Display for ParticipantIdentifier {
...
 ///
-/// Discrepancy (B): The codebase  instantiates [`Identifier`]s in three
-/// different ways: (1) as a session identifier for keygen, that are created at
-/// the start of a key generation instance; (2) as a session identifier for
-/// pre-signing, (3) as a session identifier for auxiliary information
-/// generation sessions; The paper itself only distinguishes between sessions
-/// and sub-sessions and combine (2) and (3).
+///     In fact, we think this is a minor bug in Figure 6 of the paper, since
```
Comment (indomitableSwan) [inline src/protocol.rs:362]: switch "this" to "there"

File: src/protocol.rs [hunk 6]
```diff
@@ -313,42 +324,60 @@ impl std::fmt::Display for ParticipantIdentifier {
...
 ///
-/// TODO: Discrepancy (C) needs to be further addressed by issue #218.
+///     This requires the calling application to select a protocol with
+/// appropriate trust assumptions for its deployment to ensure the resulting
+/// [`Identifier`] is unique and that all parties have the same one.
+/// Sample protocols (with varying trust models!) could include:
+///     - A trusted party randomly samples a unique identifier with
+///       `Identifier::random()` and sends it to all parties;
+///     - The participants run a Byzantine agreement protocol;
+///     - The participants use a pseudorandom correlation generator.
```
Comment (indomitableSwan) [inline src/protocol.rs:380]: ah, here it is. I think we should move this up top, to where we first start talking about session identifiers.

File: src/protocol.rs [hunk 7]
```diff
@@ -313,42 +324,60 @@ impl std::fmt::Display for ParticipantIdentifier {
...
 /// # Discrepancies with the paper with respect to session identifiers:
+/// The paper defines session identifiers, denoted `sid` and `ssid`, somewhat
+/// differently that we implement them in this codebase. We believe this
+/// implementation is functionally equivalent to what the paper describes.
+///
+/// 1. The paper incorporates many types of data into its session and
+/// sub-session identifiers, including fixed parameters, the participant set,
+/// and outputs from previous protocols; these identifiers are used both to tag
+/// messages and incorporate context into proofs. The codebase defines a single
+/// `Identifier` type that is expected to be chosen uniformly at random; this is
```
Comment (indomitableSwan) [inline src/protocol.rs:349]: this requirement seems to contradict what we say below about options for picking session identifiers

File: src/protocol.rs [hunk 8]
```diff
@@ -313,42 +324,60 @@ impl std::fmt::Display for ParticipantIdentifier {
...
+/// they correspond.
 ///
 /// # Discrepancies with the paper with respect to session identifiers:
+/// The paper defines session identifiers, denoted `sid` and `ssid`, somewhat
+/// differently that we implement them in this codebase. We believe this
+/// implementation is functionally equivalent to what the paper describes.
+///
+/// 1. The paper incorporates many types of data into its session and
+/// sub-session identifiers, including fixed parameters, the participant set,
+/// and outputs from previous protocols; these identifiers are used both to tag
```
Comment (indomitableSwan) [inline src/protocol.rs:347]: Can we be more specific about "outputs from previous protocols"? These are key- and party-specific parameters that we expect the calling application to persist across sessions, right?

File: src/protocol.rs [hunk 9]
```diff
@@ -313,42 +324,60 @@ impl std::fmt::Display for ParticipantIdentifier {
...
-/// session key and auxiliary information. The codebase does not do that and
-/// instead relies on the calling application to generate and refresh these
-/// identifiers by randomly sampling a new and unique 32 bytes identifier using
-/// `Identifier::random()`. This assumes that the participants initiating a
-/// protocol run are honestly generating globally unique identifiers and
-/// distributing them to the correct set of parties.
+/// 3. 🔒 In the paper, `ssid` is updated each time the participants run the
+/// key-refresh subprotocol.
+/// The codebase relies on the calling application to generate a new, unique
+/// `Identifier` for each new session.
```
Comment (indomitableSwan) [inline src/protocol.rs:371]: Is it worth adding a note that clarifies that the codebase expects the calling application to persist the refreshed key and party-specific parameters, independently from these session identifiers?

**Review by indomitableSwan — COMMENTED** [review]
I meant to glance at this and ended up reviewing. Lots of comments, per my usual. Thx @marsella

File: src/protocol.rs (hunk 8 above)
Comment (Ra1issa) [reply src/protocol.rs:347]: Maybe rather than say "functionally equivalent" mention that it achieves the same guarantees (i.e. global uniqueness) ?

File: src/protocol.rs (hunk 5 above)
Comment (indomitableSwan) [reply src/protocol.rs:362]: It's a bug. `ssid` contains the values you're generating by running AuxInfo.

See also Figure 3, where `sid` is passed as the session identifier input to Auxinfo (which makes way more sense).

File: src/protocol.rs (hunk 6 above)
Comment (indomitableSwan) [reply src/protocol.rs:380]: You don't need a PCG. It's just an option. 

Say you have parties P1, ..., Pn. They then want to generate 10 ECDSA keys. Can you explain how following the paper's suggestion for `sid` results in a unique session identifier for each of these 10 separate calls to KeyGen? I see only static information included.

Comment (indomitableSwan) [reply src/protocol.rs:380]: To add more on the PCG option: it's definitely overkill, but it's one way for parties to agree on values without introducing extra rounds during protocol runs. You eat the setup cost up front.

File: src/lib.rs (hunk 1 above)
Comment (marsella) [reply src/lib.rs:87]: Yeah, the weak language is because I haven't really thought about it. If you have totally independent deployments with overlapping participant IDs, maybe it's fine? But I guess even if they participants are different, it opens up the door to potential reuse-style attacks. I'll make this a stronger statement.

File: src/protocol.rs (hunk 2 above)
Comment (marsella) [reply src/protocol.rs:301]: @amaloz [just brought this up, too](https://github.com/boltlabs-inc/tss-ecdsa/pull/278#discussion_r1165920807). They do tie into message authentication, but that's a requirement for the calling application. I was going to do a separate PR that focuses on all the calling application requirements around networking (e.g. defining what type of channels they need to support and what kind of message validation they need to do).

Comment (marsella) [reply src/protocol.rs:301]: Made #294 to address this.

File: src/protocol.rs (hunk 7 above)
Comment (marsella) [reply src/protocol.rs:349]: Right. I will rephrase this to emphasize the property (global, unique) that these tags have, and not the method of choosing them.

File: src/protocol.rs (hunk 5 above)
Comment (marsella) [reply src/protocol.rs:362]: Fixed. Also moved this to a footnote because it's more of a point of interest than something the reader needs to take action on.

File: src/protocol.rs (hunk 6 above)
Comment (marsella) [reply src/protocol.rs:380]: Moved!

File: src/protocol.rs (hunk 9 above)
Comment (marsella) [reply src/protocol.rs:371]: No, I think that will be distracting here. When we actually implement key refresh, we'll write documentation on that protocol to describe the new storage requirements (e.g. throw out the old, persist the new).

File: src/protocol.rs [hunk 10]
```diff
@@ -287,7 +287,18 @@ impl ParticipantConfig {
     }
 }
 
-/// An identifier corresponding to a [`Participant`].
+/// An identifier for a [`Participant`].
+///
+/// All [`Participant`]s in a session must agree on the
+/// [`ParticipantIdentifier`]s. That is, these are not local identifiers for a
+/// single [`Participant`]; they are global identifiers for the [`Participant`]s
```
[the remark this answers is not in the document]
Comment (marsella) [reply src/protocol.rs:294]: I'm being kind of weird about this because `ParticipantIdentifier`s are unique per participant, but not necessarily per session (e.g. you can keep the same pid across keygen, auxinfo, and several presigns). I'll just state that clearly, though.

File: src/protocol.rs (hunk 8 above)
Comment (marsella) [reply src/protocol.rs:347]: Added specifics in a footnote. And rephrased (because yes, "functionally equivalent" doesn't mean anything).

**Comment by marsella** [comment]
@indomitableSwan @Ra1issa Thanks for all the comments, they are super helpful. This is ready for re-review!

File: src/protocol.rs (hunk 6 above)
Comment (Ra1issa) [reply src/protocol.rs]: What I still don't understand is what the correlation you are hoping to get out of a PCG is ? Wouldn't you get your desired outcome if the parties use a PRG with the same seed ?

Comment (indomitableSwan) [reply src/protocol.rs]: Sure - but isn't that technically still a PCG? 

I just had PCG techniques in my head when we wrote this list, because I was thinking about them for another problem. Would you prefer we remove this reference altogether? Reword in some fashion? I don't have any strong feelings about keeping/removing.

File: src/lib.rs (hunk 1 above)
Comment (indomitableSwan) [reply src/lib.rs]: I think I'm having trouble understanding the requirements here because I don't know how these identifiers relate to real-world identities. Maybe we can take another look at this language as part of #294

**Review by indomitableSwan — APPROVED** [review]
LGTM.

File: src/protocol.rs (hunk 6 above)
Comment (Ra1issa) [reply src/protocol.rs]: It is a PCG, but that seems like a very big hammer for what amounts to a PRG. Unless I am missing something. 
I'd rather if we removed this reference altogether.

Comment (marsella) [reply src/protocol.rs]: Comically oversized hammer is correct, I'll take it out.

File: src/lib.rs (hunk 1 above)
Comment (marsella) [reply src/lib.rs]: Yeah, that's fair. I added a note to that issue.
